### PR TITLE
YD-378 Implemented "last seen"

### PR DESCRIPTION
**Pull request by Bert-R**

### CODE REVIEW

**Comment by Bert-R** [comment]
The change set is complete, ready for review.
There is only one small pending change: the colors in the bar chart.

File: core/src/main/java/nu/yona/server/util/TimeUtil.java [hunk 1]
```diff
@@ -1,12 +1,10 @@
 /*******************************************************************************
- * Copyright (c) 2017 Stichting Yona Foundation
- *
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ * Copyright (c) 2017 Stichting Yona Foundation This Source Code Form is subject to the terms of the Mozilla Public License, v.
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/util/TimeUtil.java]: Actually, the formatting was better before (a full stop is missing)...

File: dbinit/src/main/liquibase/updates/changelog-0003-yd-378.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+- changeSet:
+    id: 1485291488594-1
+    author: Bert (generated and altered)
+    changes:
+    - addColumn:
+        columns:
+        - column:
+            name: app_last_opened_date
+            type: date
+            defaultValueDate: "2017-01-25"
```
Comment (aristotelos) [inline dbinit/src/main/liquibase/updates/changelog-0003-yd-378.yml]: Maybe add a comment that a value is required, and this is used to set existing records to a date.

File: adminservice/src/main/java/nu/yona/server/admin/DashboardController.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+		return "dashboard";
+	}
+
+	private List<Interval> determineIntervals(List<Integer> intervalEndOffsets)
+	{
+		Interval.setNextBegin(LocalDate.now());
+		List<Interval> intervals = intervalEndOffsets.stream().map(ieo -> Interval.nextInstance(ieo))
+				.collect(Collectors.toList());
+		intervals.add(Interval.nextInstance(LocalDate.of(2017, 1, 1)));
```
Comment (Bert-R) [inline adminservice/src/main/java/nu/yona/server/admin/DashboardController.java]: Named it LONG_AGO and set it to 1-1-2000.

File: adminservice/src/main/java/nu/yona/server/admin/DashboardController.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		return values.stream().map(v -> Math.round(((float) v) / sum * 100)).collect(Collectors.toList());
+	}
+
+	@FunctionalInterface
+	private interface Getter
+	{
+		int count(LocalDate begin, LocalDate end);
+	}
+
+	private static class Interval
```
Comment (Bert-R) [inline adminservice/src/main/java/nu/yona/server/admin/DashboardController.java]: True, but not worth the effort. It's just our internal management dashboard. Reliability of this is not our greatest concern.

File: adminservice/src/main/java/nu/yona/server/admin/DashboardController.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	{
+		final LocalDate begin;
```
Comment (Bert-R) [inline adminservice/src/main/java/nu/yona/server/admin/DashboardController.java]: Didn't know that. Done.

Comment (Bert-R) [inline adminservice/src/main/java/nu/yona/server/admin/DashboardController.java]: I've renamed it into ``HistoryInterval`` and added an explanatory comment:

```
/**
 * This interval is used to reason backward in time, so the end is more recent than the start. Note that the interval is used
 * for "between" queries, which include both start and end, so an interval with an equal start and end date covers just one
 * day.
 */
```

File: adminservice/src/main/java/nu/yona/server/admin/DashboardController.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+		Interval(LocalDate begin, LocalDate end)
+		{
+			this.begin = begin;
+			this.end = end;
+		}
+
+		static Interval nextInstance(int endOffset)
+		{
+			LocalDate end = LocalDate.now().minusDays(endOffset - 1);
```
Comment (Bert-R) [inline adminservice/src/main/java/nu/yona/server/admin/DashboardController.java]: I've renamed it into ``HistoryInterval`` and added an explanatory comment:

```
/**
 * This interval is used to reason backward in time, so the end is more recent than the start. Note that the interval is used
 * for "between" queries, which include both start and end, so an interval with an equal start and end date covers just one
 * day.
 */
```

File: adminservice/src/main/resources/templates/dashboard.html [hunk 7]
```diff
@@ -0,0 +1,145 @@
+<!DOCTYPE html>
+<html>
+<head>
+<meta charset="ISO-8859-1">
```
Comment (Bert-R) [inline adminservice/src/main/resources/templates/dashboard.html]: No idea. Eclipse generated it for me and I didn't see it. I've removed the meta tag.

File: appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy [hunk 8]
```diff
@@ -779,7 +780,72 @@ class BasicBuddyTest extends AbstractAppServiceIntegrationTest
...
+		then:
+		assertEquals(buddyBob.user.appLastOpenedDate, YonaServer.now.toLocalDate())
+		buddyBob.lastMonitoredActivityDate == null
 
+		cleanup:
+		appService.deleteUser(richard)
+		appService.deleteUser(bob)
+	}
+
+	@IgnoreRest
```
Comment (Bert-R) [inline appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy]: Indeed. Keeping my fingers crossed...

File: appservice/src/intTest/groovy/nu/yona/server/UserTest.groovy [hunk 9]
```diff
@@ -8,6 +8,7 @@ package nu.yona.server
 
 import groovy.json.*
 import nu.yona.server.test.User
+import spock.lang.IgnoreRest
```
Comment (Bert-R) [inline appservice/src/intTest/groovy/nu/yona/server/UserTest.groovy]: No. Removed it.

File: core/src/main/java/nu/yona/server/util/TimeUtil.java (hunk 1 above)
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/util/TimeUtil.java]: I agree, but that's what automatically happens when the code formatter runs. The copyright inserter formats it nicely to start with but than the formatter reformats it. Legally nothing changes, so I think it's fine.

File: dbinit/src/main/liquibase/updates/changelog-0003-yd-378.yml (hunk 2 above)
Comment (Bert-R) [inline dbinit/src/main/liquibase/updates/changelog-0003-yd-378.yml]: Done.

**Pull request merged**